### ISSUE TITLE
Goolge PaLM API models

### ISSUE THREAD

**Issue by hugoferrero** [vanna-ai/vanna#239]
Hi. I want to try vanna ai on PaLM API models (bison). Do you have any tutorial or documentation on how to set up those models on vanna?. It is not clear to me how to implement any other model if you choose "Ohter LLM" in the configuration options. Here is the code i can't figure it out how to adapt to PaLM API models:
```python 
class MyCustomLLM(VannaBase):
  def __init__(self, config=None):
    pass

  def generate_plotly_code(self, question: str = None, sql: str = None, df_metadata: str = None, **kwargs) -> str:
    # Implement here

  def generate_question(self, sql: str, **kwargs) -> str:
    # Implement here
    
  def get_followup_questions_prompt(self, question: str, question_sql_list: list, ddl_list: list, doc_list: list, **kwargs):
    # Implement here
  
  def get_sql_prompt(self, question: str, question_sql_list: list, ddl_list: list, doc_list: list, **kwargs):
    # Implement here

  def submit_prompt(self, prompt, **kwargs) -> str:
    # Implement here
            

class MyVanna(ChromaDB_VectorStore, MyCustomLLM):
    def __init__(self, config=None):
        ChromaDB_VectorStore.__init__(self, config=config)
        MyCustomLLM.__init__(self, config=config)

vn = MyVanna()
```

**Comment by hugoferrero** [comment]
Thanks for the response @andreped. I will try it, and send you feddback.

**Comment by zainhoda** [comment]
@hugoferrero if you happen to make progress on this, could you pass along your code and we can potentially integrate this into the main Vanna repo?

**Comment by andreped** [comment]
> @hugoferrero if you happen to make progress on this, could you pass along your code and we can potentially integrate this into the main Vanna repo?

I am open to drafting a PR for this today :] I can tag you in, @hugoferrero, if you wish to test it before merging.

**Comment by andreped** [comment]
I made a PR https://github.com/vanna-ai/vanna/pull/264.

It is a rather simple implementation but sadly I do not have access to Google Cloud. I am therefore dependent on some of you to test it.

To install:
```
pip install git+https://github.com/andreped/vanna.git@bison-support
pip install chromadb google-cloud-aiplatform
```

Then you should be able to initialize it with a vector DB like Chroma like so:
```
from vertexai.language_models import ChatModel
from vanna.chromadb.chromadb_vector import ChromaDB_VectorStore
from vanna.palm.palm import Palm

class MyVanna(ChromaDB_VectorStore, Palm):
        def __init__(self, config=None):
            ChromaDB_VectorStore.__init__(self, config=config)
            Palm.__init__(self, client=ChatModel("chat-bison@001"), config=config)

vn = MyVanna()

# do as you normally would with any other client(s)
# [...]
```

**Comment by yedhukr** [comment]
@andreped  I've tried the code and it's not working, it gives errors like:
_'MyVanna' object has no attribute 'temperature'_

I then tried explicitly mentioning the parameters as below:
```
def submit_prompt(self, prompt, **kwargs) -> str:
    temperature = 0.7
    max_tokens = 500
    top_p = 0.95
    top_k = 40

    params = {
        "temperature": temperature,
        "max_output_tokens": max_tokens,
        "top_p": top_p,
        "top_k": top_k,
    }

    response = self.client.send_message(prompt, **params)
    return response.text
```
        
 This gave me the error:
  _'ChatModel' object has no attribute 'send_message'_

I also attempted to specify the model more explicitly:
```
def submit_prompt(self, prompt, **kwargs) -> str:
    temperature = 0.7
    max_tokens = 500
    top_p = 0.95
    top_k = 40

    client = ChatModel("chat-bison@001")
    chat = client.start_chat()
    params = {
        "temperature": temperature,
        "max_output_tokens": max_tokens,
        "top_p": top_p,
        "top_k": top_k,
    }
    print(prompt)
    response = chat.send_message(prompt, **params)
    return response.text
```
This gave me this error:
_400 Invalid resource field value in the request._

Please help 🙏🏼

**Comment by andreped** [comment]
> @andreped I've tried the code and it's not working, it gives errors like:
> _'MyVanna' object has no attribute 'temperature'_

Hello, @yedhukr! :] Great that you were able to test the implementation!

I don't have access to Google Cloud, so I have no way of testing it. Perhaps someone could reach out, and I could borrow some API key such that I could debug this properly? Just for this PR, then the key could be rotated. @zainhoda?

**Comment by yedhukr** [comment]
Modifying the function in this manner gets it to run, but I get a response like:
_'Hi there, how can I help you today?'_

```
    def submit_prompt(self, prompt, **kwargs) -> str:
        temperature = 0.7
        max_tokens = 500
        top_p = 0.95
        top_k = 40

        chat_model = ChatModel.from_pretrained("chat-bison@001")
        chat = chat_model.start_chat()
        params = {
            "temperature": temperature,
            "max_output_tokens": max_tokens,
            "top_p": top_p,
            "top_k": top_k,
        }
        
        response = chat.send_message("{prompt}", **params)
        return response.text
```

**Comment by andreped** [comment]
> Modifying the function in this manner gets it to run, but I get a response like:
> _'Hi there, how can I help you today?'_

Which user prompt did you use? It also sounds like you are missing the system message that Vanna uses.

I think by doing this `chat = chat_model.start_chat()`, you are not setting the system message. Here you can see an example which includes a system message:
https://cloud.google.com/vertex-ai/generative-ai/docs/sdk-for-llm/sdk-use-text-models#generate-text-chat-sdk

As a test, could you try to feed the system message that the Vanna Base class uses here:
https://github.com/vanna-ai/vanna/blob/main/src/vanna/base/base.py#L448

Basically, change `chat = chat_model.start_chat()` to:
```
chat = chat_model.start_chat(
    context="The user will give you SQL and you will try to guess what the business question this query is answering. Return just the question without any additional explanation. Do not reference the table name in the question."
)
```

If that works, I think I know how to fix the issue.

**Comment by yedhukr** [comment]
Alright I made the change as you mentioned, here is the entire code for reference and the response was still incorrect:

_[{'role': 'system', 'content': 'The user provides a question and you provide SQL...;} ... {'role': 'user', 'content': 'What are the top 5 properties by sales in 2023?'}]
What are the average sales per store in each state?
What are the average sales per store in each state?
Couldn't run sql:  Execution failed on sql 'What are the average sales per store in each state?': near "What": syntax error_

```
class Palm(VannaBase):
    def __init__(self, client=None, config=None):
        VannaBase.__init__(self, config=config)

        if client is not None:
            self.client = client
            return

        # default values for params
        temperature = 0.7
        max_tokens = 500
        top_p = 0.95
        top_k = 40


    def system_message(self, message: str) -> any:
        return {"role": "system", "content": message}

    def user_message(self, message: str) -> any:
        return {"role": "user", "content": message}

    def assistant_message(self, message: str) -> any:
        return {"role": "assistant", "content": message}

    def extract_sql_query(self, text):
        """
        Extracts the first SQL statement after the word 'select', ignoring case,
        matches until the first semicolon, three backticks, or the end of the string,
        and removes three backticks if they exist in the extracted string.

        Args:
        - text (str): The string to search within for an SQL statement.

        Returns:
        - str: The first SQL statement found, with three backticks removed, or an empty string if no match is found.
        """
        # Regular expression to find 'select' (ignoring case) and capture until ';', '```', or end of string
        pattern = re.compile(r"select.*?(?:;|```|$)", re.IGNORECASE | re.DOTALL)

        match = pattern.search(text)
        if match:
            # Remove three backticks from the matched string if they exist
            return match.group(0).replace("```", "")
        else:
            return text

    def generate_sql(self, question: str, **kwargs) -> str:
        # Use the super generate_sql
        sql = super().generate_sql(question, **kwargs)

        # Replace "\_" with "_"
        sql = sql.replace("\\_", "_")

        sql = sql.replace("\\", "")

        return self.extract_sql_query(sql)

    def submit_prompt(self, prompt, **kwargs) -> str:
        temperature = 0.7
        max_tokens = 500
        top_p = 0.95
        top_k = 40

        chat_model = ChatModel.from_pretrained("chat-bison@001")
        chat = chat_model.start_chat(
            context="The user will give you SQL and you will try to guess what the business question this query is answering. Return just the question without any additional explanation. Do not reference the table name in the question."
        )
        params = {
            "temperature": temperature,
            "max_output_tokens": max_tokens,
            "top_p": top_p,
            "top_k": top_k,
        }
        
        response = chat.send_message("{prompt}", **params)
        return response.text
    
    
class MyVanna(VannaDB_VectorStore, Palm):
    def __init__(self, config=None):
        VannaDB_VectorStore.__init__(self, vanna_model="***", vanna_api_key="***", config=config)
        Palm.__init__(self, client=ChatModel("chat-bison@001"), config=config)
        print(self)

vn = MyVanna()
vn.connect_to_sqlite('database/rl_database.sqlite')
vn.ask("What are the top 5 properties by sales in 2023?")
```

**Comment by hugoferrero** [comment]
Hi Guys. Sorry, i can't lend the API KEY. I have a corporate account.

**Comment by andreped** [comment]
> Hi Guys. Sorry, i can't lend the API KEY. I have a corporate account.

No problem. I will check around if I can get a new trial. Maybe I just need to setup a new account :P
